File: disjoint-set-data-structure/DisjointSetTest.cpp
/* 
 * Disjoint-set data structure - Test suite (C++)
 * 
 * Copyright (c) 2021 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/disjoint-set-data-structure
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <random>
#include <vector>
#include "DisjointSet.hpp"

using std::size_t;


/*---- Helper definitions ----*/

class NaiveDisjointSet final {
	private: std::vector<size_t> representatives;
	
	public: explicit NaiveDisjointSet(size_t numElems) {
		for (size_t i = 0; i < numElems; i++)
			representatives.push_back(i);
	}
	
	public: size_t getNumSets() const {
		size_t result = 0;
		for (size_t i = 0; i < representatives.size(); i++) {
			if (representatives[i] == i)
				result++;
		}
		return result;
	}
	
	public: size_t getSizeOfSet(size_t elemIndex) const {
		size_t repr = representatives[elemIndex];
		size_t result = 0;
		for (size_t i = 0; i < representatives.size(); i++) {
			if (representatives[i] == repr)
				result++;
		}
		return result;
	}
	
	public: bool areInSameSet(size_t elemIndex0, size_t elemIndex1) const {
		return representatives[elemIndex0] == representatives[elemIndex1];
	}
	
	public: bool mergeSets(size_t elemIndex0, size_t elemIndex1) {
		size_t repr0 = representatives[elemIndex0];
		size_t repr1 = representatives[elemIndex1];
		for (size_t i = 0; i < representatives.size(); i++) {
			if (representatives[i] == repr1)
				representatives[i] = repr0;
		}
		return repr0 != repr1;
	}
};


// Random number generation global variables
std::default_random_engine randGen((std::random_device())());
std::uniform_real_distribution<double> realDist;


static void assertTrue(bool cond) {
	if (!cond)
		throw "Assertion error";
}


static void assertFalse(bool cond) {
	assertTrue(!cond);
}


static void assertEquals(std::int64_t expected, std::int64_t actual) {
	assertTrue(actual == expected);
}


/*---- Test suite ----*/

static void testNew() {
	DisjointSet<signed char> ds(10);
	assertEquals(10, ds.getNumSets());
	assertEquals(1, ds.getSizeOfSet(0));
	assertEquals(1, ds.getSizeOfSet(2));
	assertEquals(1, ds.getSizeOfSet(9));
	assertTrue(ds.areInSameSet(0, 0));
	assertFalse(ds.areInSameSet(0, 1));
	assertFalse(ds.areInSameSet(9, 3));
	ds.checkStructure();
}


static void testMerge() {
	DisjointSet<std::uint16_t> ds(10);
	assertTrue(ds.mergeSets(0, 1));
	ds.checkStructure();
	assertEquals(9, ds.getNumSets());
	assertTrue(ds.areInSameSet(0, 1));
	
	assertTrue(ds.mergeSets(2, 3));
	ds.checkStructure();
	assertEquals(8, ds.getNumSets());
	assertTrue(ds.areInSameSet(2, 3));
	
	assertFalse(ds.mergeSets(2, 3));
	ds.checkStructure();
	assertEquals(8, ds.getNumSets());
	assertFalse(ds.areInSameSet(0, 2));
	
	assertTrue(ds.mergeSets(0, 3));
	ds.checkStructure();
	assertEquals(7, ds.getNumSets());
	assertTrue(ds.areInSameSet(0, 2));
	assertTrue(ds.areInSameSet(3, 0));
	assertTrue(ds.areInSameSet(1, 3));
}


static void testBigMerge() {
	const int MAX_RANK = 20;
	const long TRIALS = 10000;
	
	const size_t numElems = static_cast<size_t>(1) << MAX_RANK;  // Grows exponentially
	DisjointSet<std::uint32_t> ds(numElems);
	for (int level = 0; level < MAX_RANK; level++) {
		size_t mergeStep = static_cast<size_t>(1) << level;
		size_t incrStep = mergeStep * 2;
		for (size_t i = 0; i < numElems; i += incrStep) {
			assertFalse(ds.areInSameSet(i, i + mergeStep));
			assertTrue(ds.mergeSets(i, i + mergeStep));
		}
		// Now we have a bunch of sets of size 2^(level+1)
		
		// Do random tests
		size_t mask = -incrStep;  // 0b11...100...00
		for (long i = 0; i < TRIALS; i++) {
			size_t j = (std::uniform_int_distribution<size_t>(0, numElems - 1))(randGen);
			size_t k = (std::uniform_int_distribution<size_t>(0, numElems - 1))(randGen);
			bool expect = (j & mask) == (k & mask);
			assertTrue(ds.areInSameSet(j, k) == expect);
		}
	}
}


static void testAgainstNaiveRandomly() {
	const long TRIALS = 1000;
	const long ITERATIONS = 3000;
	const size_t NUM_ELEMS = 300;
	
	for (long i = 0; i < TRIALS; i++) {
		NaiveDisjointSet nds(NUM_ELEMS);
		DisjointSet<size_t> ds(NUM_ELEMS);
		for (long j = 0; j < ITERATIONS; j++) {
			size_t k = (std::uniform_int_distribution<size_t>(0, NUM_ELEMS - 1))(randGen);
			size_t l = (std::uniform_int_distribution<size_t>(0, NUM_ELEMS - 1))(randGen);
			assertTrue(ds.getSizeOfSet(k) == nds.getSizeOfSet(k));
			assertTrue(ds.areInSameSet(k, l) == nds.areInSameSet(k, l));
			if (realDist(randGen) < 0.1)
				assertTrue(ds.mergeSets(k, l) == nds.mergeSets(k, l));
			assertEquals(nds.getNumSets(), ds.getNumSets());
			if (realDist(randGen) < 0.001)
				ds.checkStructure();
		}
		ds.checkStructure();
	}
}


/*---- Main runner ----*/

int main() {
	try {
		testNew();
		testMerge();
		testBigMerge();
		testAgainstNaiveRandomly();
		
		std::cerr << "Test passed" << std::endl;
		return EXIT_SUCCESS;
	} catch (const char *msg) {
		std::cerr << msg << std::endl;
		return EXIT_FAILURE;
	}
}

File: disjoint-set-data-structure/disjointset-test.rs
/* 
 * Disjoint-set data structure - Test suite (Rust)
 * 
 * Copyright (c) 2019 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/disjoint-set-data-structure
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

extern crate rand;
use rand::Rng;
use rand::distributions::IndependentSample;
mod disjointset;
use disjointset::DisjointSet;


/*---- Main runner ----*/

fn main() {
	test_new();
	test_merge();
	test_big_merge();
	test_against_naive_randomly();
	println!("Test passed");
}



/*---- Test suite ----*/

fn test_new() {
	let mut ds = DisjointSet::new(10);
	assert_eq!(10, ds.number_of_sets());
	assert_eq!(1, ds.get_size_of_set(0));
	assert_eq!(1, ds.get_size_of_set(2));
	assert_eq!(1, ds.get_size_of_set(9));
	assert_eq!(true, ds.are_in_same_set(0, 0));
	assert_eq!(false, ds.are_in_same_set(0, 1));
	assert_eq!(false, ds.are_in_same_set(9, 3));
}


fn test_merge() {
	let mut ds = DisjointSet::new(10);
	assert_eq!(true, ds.merge_sets(0, 1));
	ds.check_structure();
	assert_eq!(9, ds.number_of_sets());
	assert_eq!(true, ds.are_in_same_set(0, 1));
	
	assert_eq!(true, ds.merge_sets(2, 3));
	ds.check_structure();
	assert_eq!(8, ds.number_of_sets());
	assert_eq!(true, ds.are_in_same_set(2, 3));
	
	assert_eq!(false, ds.merge_sets(2, 3));
	ds.check_structure();
	assert_eq!(8, ds.number_of_sets());
	assert_eq!(false, ds.are_in_same_set(0, 2));
	
	assert_eq!(true, ds.merge_sets(0, 3));
	ds.check_structure();
	assert_eq!(7, ds.number_of_sets());
	assert_eq!(true, ds.are_in_same_set(0, 2));
	assert_eq!(true, ds.are_in_same_set(3, 0));
	assert_eq!(true, ds.are_in_same_set(1, 3));
}


fn test_big_merge() {
	let maxrank: i8 = 20;
	let trials: i32 = 10000;
	
	let numelems: usize = 1 << maxrank;  // Grows exponentially
	let mut ds = DisjointSet::new(numelems);
	let rng = &mut rand::thread_rng();
	let range = rand::distributions::range::Range::new(0, numelems);
	for level in 0 .. maxrank {
		let mergestep: usize = 1 << level;
		let incrstep: usize = mergestep * 2;
		let mut i: usize = 0;
		while i < numelems {
			assert_eq!(false, ds.are_in_same_set(i, i + mergestep));
			assert_eq!(true, ds.merge_sets(i, i + mergestep));
			i += incrstep;
		}
		// Now we have a bunch of sets of size 2^(level+1)
		
		// Do random tests
		let mask: usize = incrstep.wrapping_neg();  // 0b11...100...00
		for _ in 0 .. trials {
			let j: usize = range.ind_sample(rng);
			let k: usize = range.ind_sample(rng);
			let expect: bool = (j & mask) == (k & mask);
			assert_eq!(expect, ds.are_in_same_set(j, k));
		}
	}
}


fn test_against_naive_randomly() {
	let trials: i32 = 1000;
	let iterations: i32 = 3000;
	let numelems: usize = 300;
	
	let rng = &mut rand::thread_rng();
	let range = rand::distributions::range::Range::new(0, numelems);
	for _ in 0 .. trials {
		let mut nds = NaiveDisjointSet::new(numelems);
		let mut ds = DisjointSet::new(numelems);
		for _ in 0 .. iterations {
			let i: usize = range.ind_sample(rng);
			let j: usize = range.ind_sample(rng);
			assert_eq!(nds.get_size_of_set(i), ds.get_size_of_set(i));
			assert_eq!(nds.are_in_same_set(i, j), ds.are_in_same_set(i, j));
			if rng.next_f64() < 0.1 {
				assert_eq!(nds.merge_sets(i, j), ds.merge_sets(i, j));
			}
			assert_eq!(nds.number_of_sets(), ds.number_of_sets());
			if rng.next_f64() < 0.001 {
				ds.check_structure();
			}
		}
		ds.check_structure();
	}
}



/*---- Helper definitions ----*/

struct NaiveDisjointSet {
	representatives: Vec<usize>,
}


impl NaiveDisjointSet {
	
	fn new(numelems: usize) -> Self {
		Self { representatives: (0usize .. numelems).collect() }
	}
	
	
	fn number_of_sets(&self) -> usize {
		self.representatives.iter().enumerate()
			.filter(|&(i, &repr)| repr == i).count()
	}
	
	
	fn get_size_of_set(&self, elemindex: usize) -> usize {
		let repr: usize = self.representatives[elemindex];
		self.representatives.iter().filter(|&r| *r == repr).count()
	}
	
	
	fn are_in_same_set(&self, elemindex0: usize, elemindex1: usize) -> bool {
		self.representatives[elemindex0] == self.representatives[elemindex1]
	}
	
	
	fn merge_sets(&mut self, elemindex0: usize, elemindex1: usize) -> bool {
		let repr0: usize = self.representatives[elemindex0];
		let repr1: usize = self.representatives[elemindex1];
		for c in self.representatives.iter_mut() {
			if *c == repr1 {
				*c = repr0;
			}
		}
		repr0 != repr1
	}
	
}
